Persistent Node Names

bleskes opened the pull request:
With #19140 we started persisting the node ID across node restarts. Now that we have a "stable" anchor, we can use it to generate a stable default node name and make it easier to track nodes over a restarts. Sadly, this means we will not have those random fun Marvel characters but we feel this is the right tradeoff.

On the implementation side, this requires a bit of juggling because we now need to read the node id from disk before we can log as the node node is part of each log message. The PR move the initialization of NodeEnvironment as high up in the starting sequence as possible, with only one logging message before it to indicate we are initializing. Things look now like this:

```
[2016-07-15 19:38:39,742][INFO ][node                     ] [_unset_] initializing ...
[2016-07-15 19:38:39,826][INFO ][node                     ] [aAmiW40] node name set to [aAmiW40] by default. set the [node.name] settings to change it
[2016-07-15 19:38:39,829][INFO ][env                      ] [aAmiW40] using [1] data paths, mounts [[ /(/dev/disk1)]], net usable_space [5.5gb], net total_space [232.6gb], spins? [unknown], types [hfs]
[2016-07-15 19:38:39,830][INFO ][env                      ] [aAmiW40] heap size [1.9gb], compressed ordinary object pointers [true]
[2016-07-15 19:38:39,837][INFO ][node                     ] [aAmiW40] version[5.0.0-alpha5-SNAPSHOT], pid[46048], build[473d3c0/2016-07-15T17:38:06.771Z], OS[Mac OS X/10.11.5/x86_64], JVM[Oracle Corporation/Java HotSpot(TM) 64-Bit Server VM/1.8.0_51/25.51-b03]
[2016-07-15 19:38:40,980][INFO ][plugins                  ] [aAmiW40] modules [percolator, lang-mustache, lang-painless, reindex, aggs-matrix-stats, lang-expression, ingest-common, lang-groovy, transport-netty], plugins []
[2016-07-15 19:38:43,218][INFO ][node                     ] [aAmiW40] initialized
```

Needless to say, settings `node.name` explicitly still works as before.

The PR also contains some clean ups to the relationship between Environment, Settings and Plugins. The previous code suggested the path related settings could be changed after the initial Environment was changed. This did not have any effect as the security manager already locked things down.

jasontedor commented on core/src/main/java/org/elasticsearch/env/Environment.java:
```diff
@@ -332,4 +339,26 @@ public Path tmpFile() {
+    }
+
+    private static void assertEquals(Object actual, Object expected, String name) {
+        assert Objects.deepEquals(actual, expected) : "actual " + name + " [" + actual + "] is different than [ " + expected + "]";
```
Are we sure that the array-valued attributes (e.g., `dataWithClusterFiles`) will always be in the same order?

bleskes replied:
That's a good one. I checked and putArray & getArray maintain order, so I believe we're good.

jasontedor commented on core/src/main/java/org/elasticsearch/env/Environment.java:
```diff
@@ -332,4 +339,26 @@ public Path tmpFile() {
+     * assserts that the two environments are equivalent for all things the environment cares about (i.e., all but the setting
+     * object which may contain different setting)
+     */
+    public static void assertEquivalent(Environment actual, Environment expected) {
```
This is only used in `Node`, and only relies on public methods on `Environment`. I'd rather see this as a private method in `Node`.

bleskes replied:
sure. Will move - does mean the name will be longer.

bleskes replied:
on second thought ( after I already moved it :)) -  have it on node means that if someone adds a path or something similar to Environment she's less likely to adapt this method. Feels to me like it should better stay in Environment.

jasontedor replied:
I'm not convinced that keeping it here has any impact on whether or not when someone modifies `Environment` they are more or less likely to adapt this method. If that's a concern, we need something stronger than pushing it into `Environment` and increasing the public API. I'm still of the mind that this should be on `Node` and private.

bleskes replied:
It only feels natural to me to have the logic comparing two environments to be in the environment class. Makes it easier to see it's correct and that it doesn't miss anything.

> we need something stronger 

I'm open for suggestions.

jasontedor replied:
> It only feels natural to me to have the logic comparing two environments to be in the environment class.

Since its only use right now is comparing two environments within `Node`, I don't agree.

bleskes replied:
> Since its only use right now is comparing two environments within Node, I don't agree.

That's clear and noted. You said that to keep the check here for the reason if we add more fields we won't forget them in the assert you would like more checks. Can you make a suggestion as to what to add?

To be honest, I added these because I'm conflicted between two things:

1) I don't want the Environment to expose a different settings than the one the node has.
2) I don't want the setting in the Environment to reflect something else then was actually used in the security manager, for example. 

I chose to reinitialize the Environment with the new settings and make sure nothing of the things that matter changed (it currently can't so I added an assertion).

I will try to another route and remote the Environment.getSettings() method, which means all of this goes away. I will do it in another PR

bleskes replied:
> I will try to another route and remote the Environment.getSettings() method, which means all of this goes away. I will do it in another PR

OK. I looked at this more carefully and this will be a good but big change (still a good one though). I rather not embark on it now. 

This brings me back to the question:

> Can you make a suggestion as to what to add?

jasontedor replied:
> Can you make a suggestion as to what to add?

Let's recap. I want the method moved, I don't want the public surface of `Environment` increased. You don't want to move it because you think it makes it more likely that if someone modifies `Environment`, they will modify this method. I do not agree with that, I do not think that it makes it more likely. I am saying that if you feel that keeping this method on `Environment` for the reason of making it more likely that it's modified in conjunction with `Environment` is a good enough reason to increase the public surface of `Environment`, then I want something stronger. So, to be clear, I don't even agree with the premise.

> I will try to another route and remote the Environment.getSettings() method, which means all of this goes away. I will do it in another PR

That's fine.

jasontedor replied:
Okay. If you have a plan to remove settings from `Environment`, then I think we can do this? Leave the method here. When you remove settings, change the method to be the implementation of `equals` (and add an implementation of `hashCode`). What do you think?

jasontedor replied:
That is, it will be a proper equals, and you can just assert in `Node` that the two environments are equal.

jasontedor replied:
Or am I confused what your plan is?

bleskes replied:
The idea is to indeed remove the setting field from environment and stop using it as a vehicle for settings which  are also available via other places (Node) and may or may not be the same. If we do that, then I don't need to reinitialize the environment and can use a single one, which make the assertion not needed and it will go away - no need for `equal`.

That said, the change to environment will require some other changes which may be resisted - so no promises :)

jasontedor replied:
> the change to environment will require some other changes which may be resisted

Do you have a rough idea of what these changes are?

jasontedor replied:
> Do you have a rough idea of what these changes are?

I discussed this with @bleskes via another channel and he has a good plan that I can get behind.

jasontedor commented on core/src/main/java/org/elasticsearch/node/Node.java:
```diff
@@ -188,53 +197,82 @@
+
+            tmpSettings = TribeService.processSettings(tmpSettings);
+
+            // create the node environment as soon as possible, to recovery the node id and enable logging
```
Nit: `recovery` -> `recover`

bleskes replied:
check

jasontedor commented on core/src/main/java/org/elasticsearch/node/Node.java:
```diff
@@ -188,53 +197,82 @@
+                tmpSettings = addNodeNameIfNeeded(tmpSettings, nodeEnvironment.nodeId());
+            ESLogger logger = Loggers.getLogger(Node.class, tmpSettings);
+            if (hadPredefinedNodeName == false) {
+                logger.info("node name set to [{}] by default. set the [{}] settings to change it",
```
Is this log message really necessary at the info level? Either way, there are some grammar issues (e.g., `settings`) but I think it reads more simply as `"node name [{}] derived from node ID; set [{}] to override"`

bleskes replied:
I think it's good to increase the visibility and encourage people to set their node names. I personally don't mind much to move it to debug.  Since @clintongormley asked for it, I'm pulling him in to get his opinion.

clintongormley replied:
I'd like to keep the message, but I like @jasontedor 's version

jasontedor commented on core/src/main/java/org/elasticsearch/node/Node.java:
```diff
@@ -188,53 +197,82 @@
+        {
+            // use temp logger just to say we are starting. we can't use it later on because the node name might not be set
+            ESLogger logger = Loggers.getLogger(Node.class, NODE_NAME_SETTING.get(environment.settings()));
+            logger.info("initializing ...");
```
Why log this before setting the node name, it can wait a few more lines?

jasontedor replied:
I see, if it waits it will not be the first log line on info level. 😞

jasontedor replied:
I do not like the `_unset_` that comes out of this log message. Now there are three prefixes: empty prefix (e.g., from the jar hell check log messages when debug logging is enabled, which occur before the node is constructed, and then also occur for module and plugin loading), the node name prefix, and this `_unset_` that appears on a single line. We've accepted the empty prefix in the past, and I think we should just do so here?

bleskes replied:
> empty prefix (e.g., from the jar hell check log messages when debug logging is enabled

Sure. I can change that based on what we decide.

> We've accepted the empty prefix in the past, and I think we should just do so here?

To me an empty `[]` feels like a bug. I don't mind much. Again asking @clintongormley for opinion as he was in favor of the some like the unset (actually he said `unnamed`, but I forgot :))

clintongormley replied:
I'm good with dropping the `_unset_`

bleskes replied:
`_unset_` be gone.

jasontedor commented on core/src/main/java/org/elasticsearch/node/Node.java:
```diff
@@ -188,53 +197,82 @@
+
+            // create the node environment as soon as possible, to recovery the node id and enable logging
+            try {
+                nodeEnvironment = new NodeEnvironment(tmpSettings,environment);
```
Whitespace between `tmpSettings,environment` -> `tmpSettings, environment`

bleskes replied:
I stopped formatting on commit. This is the result :( ... fixing.

jasontedor replied:
I'm filled with joy that you stopped formatting on commit, but you can still use the IDE to format blocks of code that you edit. 😇

jasontedor commented on core/src/main/java/org/elasticsearch/cluster/node/DiscoveryNode.java:
```diff
@@ -204,7 +203,11 @@ public static DiscoveryNode createLocal(Settings settings, TransportAddress publ
-        return new DiscoveryNode(Node.NODE_NAME_SETTING.get(settings), nodeIdSupplier, publishAddress,
+        String nodeName = Node.NODE_NAME_SETTING.get(settings);
+        if (nodeName.isEmpty()) {
+            nodeName = nodeId.substring(0, 7);
```
`assert nodeId.length >= 7`?

bleskes replied:
this is actually a left over. Reverted.

jasontedor commented on core/src/main/java/org/elasticsearch/common/settings/Setting.java:
```diff
@@ -503,7 +503,11 @@ public void apply(T value, Settings current, Settings previous) {
+        return simpleString(key, "", properties);
+    }
+
+    public static Setting<String> simpleString(String key, String defaultValue, Property... properties) {
```
I'd rather see the `_unset_` thing removed and then this method is not needed.

bleskes replied:
sure. let's see how it goes.

jasontedor commented on core/src/main/java/org/elasticsearch/node/internal/InternalSettingsPreparer.java:
```diff
@@ -167,42 +161,6 @@ private static void finalizeSettings(Settings.Builder output, Terminal terminal,
-                throw new RuntimeException("Failed to load custom names.txt from " + namesPath, e);
-            }
-        } else {
-            input = InternalSettingsPreparer.class.getResourceAsStream("/config/names.txt");
```
:cry:

jasontedor commented on core/src/main/java/org/elasticsearch/node/Node.java:
```diff
@@ -316,20 +348,23 @@ protected Node(Environment tmpEnv, Collection<Class<? extends Plugin>> classpath
 
-            client.intialize(injector.getInstance(new Key<Map<GenericAction, TransportAction>>() {}));
+            client.intialize(injector.getInstance(new Key<Map<GenericAction, TransportAction>>() {
+            }));
```
The line breakage is awkward here.

bleskes replied:
removed

bleskes commented:
@jasontedor thx. I addressed all comments. Now the logs like so:

```
[elasticsearch] [2016-07-19 16:23:00,401][INFO ][node                     ] [] initializing ...
[elasticsearch] [2016-07-19 16:23:00,479][INFO ][node                     ] [e0VVSNc] node name [e0VVSNc] derived from node ID; set [node.name] to override
[elasticsearch] [2016-07-19 16:23:00,500][INFO ][env                      ] [e0VVSNc] using [1] data paths, mounts [[/ (/dev/disk1)]], net usable_space [6.7gb], net total_space [232.6gb], spins? [unknown], types [hfs]
[elasticsearch] [2016-07-19 16:23:00,500][INFO ][env                      ] [e0VVSNc] heap size [1.9gb], compressed ordinary object pointers [true]
```

jasontedor commented:
> `[elasticsearch] [2016-07-19 16:23:00,500][INFO ][env                      ] [e0VVSNc] using [1] data paths, mounts [[/ (/dev/disk1)]], net usable_space [6.7gb], net total_space [232.6gb], spins? [unknown], types [hfs]`

Getting little a tight there, be careful, you might run into strange issues with `git`. :wink:

jasontedor commented on core/src/main/java/org/elasticsearch/node/Node.java:
```diff
@@ -170,7 +170,16 @@
+     * Adds a default node name to the given setting, if it doesn't already exist
+     * @return the given setting if node name is already set, or a new copy with a default node name set.
+     */
+    public static final Settings addNodeNameIfNeeded(Settings settings, final String nodeId) {
```
This is such a hack, but it's all that we can do. :cry:

jasontedor commented:
@bleskes Can you resolve the merge conflicts and I will take a final look?

bleskes commented:
@jasontedor I updated the pr resolving the merge.

jasontedor commented:
Punch it.
